src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -817,9 +817,6 @@ protected override void Dispose(bool disposing)
             catch (Exception e)
             {
                 Log.Error("Exception caught while disposing a reader", e, Connector.Id);
-            }
-            finally
```

Comment:
@vonzshik I'm not sure how this change is related to the bug fix - is it important to set the reader state to Disposed in Close instead of here?

Reply:
It's not really related to the bug fix, but it's still a bug. If on reader's dispose we also close the connection, there will be a bug, when we change the state of the reader after the connection is returned to the pool.

Reply:
Makes sense. Ideally we'd have a test for this case too, if that's not too much work :)